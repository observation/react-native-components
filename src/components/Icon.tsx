import React from 'react'

import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'

import Icons, { IconName } from '../lib/Icons'
import theme from '../styles/theme'

export type IconStyleProp = 'light' | 'solid'

type IconAppearanceProps = {
  style?: IconStyleProp
  color?: string
  size?: number
  testID?: string
}

type IconProps = IconAppearanceProps & {
  name: IconName
}

export const Icon = ({ name, color, size, testID, style }: IconProps): JSX.Element => {
  const iconStyle = style ?? 'light'
  const icon = iconStyle === 'light' ? Icons[name].light : Icons[name].solid
  const iconColor = color ?? theme.color.primary
  const iconSize = size ?? theme.icon.size.large

  return <FontAwesomeIcon icon={icon} color={iconColor} size={iconSize} testID={testID} />
}
