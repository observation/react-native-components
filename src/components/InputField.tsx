import React, { RefAttributes, useState } from 'react'
import { Platform, TextInput, Text, View, StyleSheet, ViewStyle, TextInputProps, StyleProp } from 'react-native'

import { IconError } from './Icon'
import IconText from './IconText'
import font from '../styles/font'
import textStyle from '../styles/text'
import theme from '../styles/theme'

type Props = {
  containerStyle?: StyleProp<ViewStyle>
  inputProps?: TextInputProps & RefAttributes<TextInput>
  inputStyle?: StyleProp<ViewStyle>
  label?: string
  rightIcon?: JSX.Element
  description?: string
  errorMessage?: string
}

const InputField = ({ containerStyle, inputProps, inputStyle, label, rightIcon, description, errorMessage }: Props) => {
  const [isFocused, setIsFocused] = useState(false)

  // Set lineHeight to 0 to fix vertical alignment of the input text on iOS
  // Do this only for iOS, as setting it to 0 on Android results in input text being invisible
  const fixInputStyle = Platform.OS === 'ios' ? { lineHeight: 0 } : {}

  const hasErrors = !!errorMessage
  const borderColor = theme.getBorderColor({ isFocused, hasErrors })

  return (
    <View style={[styles.containerStyle, containerStyle]}>
      {label && (
        <View>
          <Text style={styles.labelStyle}>{label}</Text>
        </View>
      )}
      <View style={{ flexDirection: 'row' }}>
        <TextInput
          style={[{ borderColor }, styles.inputStyle, inputStyle, fixInputStyle]}
          {...inputProps}
          autoCapitalize="none"
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          underlineColorAndroid="transparent"
          placeholderTextColor={theme.color.placeholder}
        />
        {rightIcon && <View style={styles.rightIcon}>{rightIcon}</View>}
      </View>
      {errorMessage && (
        <IconText
          icon={<IconError size={theme.icon.size.small} />}
          text={errorMessage}
          style={{
            textStyle: styles.errorStyle,
            containerStyle: { marginTop: theme.margin.half },
          }}
        />
      )}
      {description && <Text style={{ marginTop: theme.margin.half, ...styles.descriptionStyle }}>{description}</Text>}
    </View>
  )
}

export default InputField

const styles = StyleSheet.create({
  containerStyle: {
    flexDirection: 'column',
  },
  labelStyle: {
    ...textStyle.inputLabel,
    marginBottom: theme.margin.half,
  },
  inputStyle: {
    ...theme.input,
    ...textStyle.input,
  },
  rightIcon: {
    ...theme.absoluteRight,
    justifyContent: 'center',
  },
  errorStyle: {
    ...font.small,
    color: theme.color.error,
  },
  descriptionStyle: {
    ...font.small,
    color: theme.color.greyDark,
  },
})
