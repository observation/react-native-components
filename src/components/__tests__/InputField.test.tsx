import React from 'react'

import { act, render } from '@testing-library/react-native'

import { Icon } from '../Icon'
import InputField from '../InputField'

describe('InputField', () => {
  describe('Rendering', () => {
    test('Normal', () => {
      // GIVEN
      const { toJSON } = render(<InputField />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With focus', async () => {
      // GIVEN
      const { toJSON, getByTestId } = render(<InputField inputProps={{ testID: 'test-id' }} />)

      // WHEN
      act(() => getByTestId('test-id').props.onFocus())

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('Without focus', async () => {
      // GIVEN
      const { toJSON, getByTestId } = render(<InputField inputProps={{ testID: 'test-id' }} />)

      // WHEN
      getByTestId('test-id').props.onBlur()

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With a label', async () => {
      // GIVEN
      const { toJSON } = render(<InputField label="The label" />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With a right icon', async () => {
      // GIVEN
      const { toJSON } = render(<InputField rightIcon={<Icon name="info-circle" />} />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With an error message', async () => {
      // GIVEN
      const { toJSON } = render(<InputField errorMessage={'The error message'} />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With focus and an error message', async () => {
      // GIVEN
      const { toJSON, getByTestId } = render(<InputField inputProps={{ testID: 'test-id' }} />)

      // WHEN
      act(() => getByTestId('test-id').props.onFocus())

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })

    test('With a description', () => {
      // GIVEN
      const { toJSON } = render(<InputField description="The description" />)

      // THEN
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
