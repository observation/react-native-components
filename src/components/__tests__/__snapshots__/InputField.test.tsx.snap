// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField Rendering Normal 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#E6E6E6",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`InputField Rendering With a description 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#E6E6E6",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      {
        "color": "#666666",
        "fontFamily": "Ubuntu",
        "fontSize": 12,
        "fontStyle": "normal",
        "fontWeight": "normal",
        "lineHeight": 17,
        "marginTop": 8,
      }
    }
  >
    The description
  </Text>
</View>
`;

exports[`InputField Rendering With a label 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View>
    <Text
      style={
        {
          "color": "#666666",
          "fontFamily": "Ubuntu",
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": "bold",
          "lineHeight": 17,
          "marginBottom": 8,
        }
      }
    >
      The label
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#E6E6E6",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`InputField Rendering With a right icon 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#E6E6E6",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
    <View
      style={
        {
          "bottom": 0,
          "justifyContent": "center",
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Icon
        color="#0066B1"
        name="circle-info"
        size={16}
        type="light"
      />
    </View>
  </View>
</View>
`;

exports[`InputField Rendering With an error message 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#EA554B",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginTop": 8,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "justifyContent": "center",
            "marginRight": 8,
          },
          undefined,
        ]
      }
    >
      <Icon
        color="#EA554B"
        name="triangle-exclamation"
        size={12}
        type="light"
      />
    </View>
    <Text
      style={
        [
          {
            "flexShrink": 1,
          },
          {
            "color": "#EA554B",
            "fontFamily": "Ubuntu",
            "fontSize": 12,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 17,
          },
        ]
      }
    >
      The error message
    </Text>
  </View>
</View>
`;

exports[`InputField Rendering With focus 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#66A3D0",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      testID="test-id"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`InputField Rendering With focus and an error message 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#66A3D0",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      testID="test-id"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`InputField Rendering Without focus 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#939393"
      style={
        [
          {
            "borderColor": "#E6E6E6",
          },
          {
            "borderRadius": 4,
            "borderWidth": 2,
            "color": "#212121",
            "flex": 1,
            "fontFamily": "Ubuntu",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
            "lineHeight": 26,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingLeft": 8,
            "paddingRight": 32,
          },
          undefined,
          {
            "lineHeight": 0,
          },
        ]
      }
      testID="test-id"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
